features/mbedtls/targets/TARGET_STM/aes_alt.c

```diff
@@ -0,0 +1,495 @@
+{
+    switch( keybits )
+    {
+        case 128:
+          ctx->nr = 10;
+          memcpy(ctx->aes_enc_key, key, 16);
+          hcryp_aes.Init.KeySize = CRYP_KEYSIZE_128B;
+          break;
+        case 192:
+          ctx->nr = 12;
+          memcpy(ctx->aes_enc_key, key, 24);
+          hcryp_aes.Init.KeySize = CRYP_KEYSIZE_192B;
+          break;
+        case 256:
+          ctx->nr = 14;
+          memcpy(ctx->aes_enc_key, key, 32);
+          hcryp_aes.Init.KeySize = CRYP_KEYSIZE_256B;
+          break;
+       default : return( MBEDTLS_ERR_AES_INVALID_KEY_LENGTH );
+    }
+
+    /* Deinitializes the CRYP peripheral */
+    HAL_CRYP_DeInit(&hcryp_aes);
+
+    hcryp_aes.Init.DataType = CRYP_DATATYPE_8B;
+
+    hcryp_aes.Instance = CRYP;
+    /* Enable CRYP clock */
+    __HAL_RCC_CRYP_CLK_ENABLE();
+
+    hcryp_aes.Init.pKey = ctx->aes_enc_key;
+    HAL_CRYP_Init(&hcryp_aes);
+    
+    return(0);
+}
+#endif /* MBEDTLS_AES_SETKEY_END_ALT */
+
+/**
+ * \brief          AES key schedule (decryption)
+ *
+ * \param ctx      AES context to be initialized
+ * \param key      decryption key
+ * \param keybits  must be 128, 192 or 256
+ *
+ * \return         0 if successful, or MBEDTLS_ERR_AES_INVALID_KEY_LENGTH
+ */
+#if defined(MBEDTLS_AES_SETKEY_DEC_ALT)
+int mbedtls_aes_setkey_dec( mbedtls_aes_context *ctx, const unsigned char *key,
+                            unsigned int keybits )
+{
+    switch( keybits )
+    {
+      case 128:
+        ctx->nr = 10;
+        memcpy(ctx->aes_dec_key, key, 16);
+        hcryp_aes.Init.KeySize = CRYP_KEYSIZE_128B;
+        break;
+      case 192:
+        ctx->nr = 12;
+        memcpy(ctx->aes_dec_key, key, 24);
+        hcryp_aes.Init.KeySize = CRYP_KEYSIZE_192B;
+        break;
+      case 256:
+        ctx->nr = 14;
+        memcpy(ctx->aes_dec_key, key, 32);
+        hcryp_aes.Init.KeySize = CRYP_KEYSIZE_256B;
+        break;
+      default : return( MBEDTLS_ERR_AES_INVALID_KEY_LENGTH );
+    }
+
+    /* Deinitializes the CRYP peripheral */
+    HAL_CRYP_DeInit(&hcryp_aes);
+    /* Enable CRYP clock */
+    hcryp_aes.Init.DataType = CRYP_DATATYPE_8B;
+
+    hcryp_aes.Instance = CRYP;
+    /* Enable CRYP clock */
+    __HAL_RCC_CRYP_CLK_ENABLE();
+
+    hcryp_aes.Init.pKey = ctx->aes_dec_key;
+    
+    HAL_CRYP_Init(&hcryp_aes);
+    
+    return( 0 );
+}
+#endif /* MBEDTLS_AES_SETKEY_DEC_ALT */
+
+/**
+ * \brief          AES-ECB block encryption/decryption
+ *
+ * \param ctx      AES context
+ * \param mode     MBEDTLS_AES_ENCRYPT or MBEDTLS_AES_DECRYPT
+ * \param input    16-byte input block
+ * \param output   16-byte output block
+ *
+ * \return         0 if successful
+ */
+int mbedtls_aes_crypt_ecb( mbedtls_aes_context *ctx,
+                    int mode,
+                    const unsigned char input[16],
+                    unsigned char output[16] )
+{
+
+    if ((ctx->nr != 10) && (ctx->nr != 12) && (ctx->nr != 14))
+        return(MBEDTLS_ERR_AES_INVALID_KEY_LENGTH);
```

Comment:
Is this check necessary here? It seems that the functions that set the encryption and decryption key already do this checks. Also, please note that this is checking whether the number of rounds is valid and not the key length. Therefor, if there was a failure, the return code does not relate to the actual problem found.